### PR TITLE
In-memory engine: fix repeatable schedule delete-range for the same range

### LINKED ISSUE #17259: In-memory Engine: delete range task has been scheduled multiple times
## background
Cache range can be decoupling with region range due to batch split. When we are to evict a range where it is a sub range of a cached range, we split the cached range so that a sub range will not de-cache a large range.  But we need to clean the data of the evicted range and this can not be done immediately due to ongoing snapshot. So we have `historical_ranges` to record the parent range and `ranges_being_deleted` to record the evicted range.
For example: 
[k1, k10] is a cached range and we evict [k3, k6]
so we have `historical_ranges` [k1, k10] and `ranges_being_deleted` [k3, k6].
When we dropped the last snapshot of [k1, k10], we will remove [k1, k10] from `ranges_being_deleted` and schedule delte range tasks for ranges in `ranges_being_deleted` that are not overlapped with any range in `historical_ranges`.
We cannot use the removed range from `historical_ranges` to select ranges in `ranges_being_deleted` directly as there may be some overlapped ranges in `historical_ranges`.
For example:
go further from the above case:
`cached_ranges`: [k1, k3], [k6, k10]
`historical_ranges`: [k1, k10]
`ranges_being_deleted`: [k3, k6],
If we evict [k6, k8], range mta becomes:
`cached_ranges`: [k1, k3], [k8, k10]
`historical_ranges`: [k1, k10], [k6, k10]
`ranges_being_deleted`: [k3, k6], [k6, k8]
The last snapshot drop of [k6, k10] cannot make [k6, k8] cleanup due to onging snashot of [k1, k10].

## what goes wrong
This process does not consider loading range that has been canceled.

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #17259

What's Changed:

```commit-message
 fix repeatable schedule delete-range for the same range
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/range_cache_memory_engine/src/background.rs [hunk 1]
```diff
@@ -1141,11 +1151,31 @@ impl Runnable for DeleteRangeRunner {
...
+                                    if !scheduled {
+                                        panic!(
+                                            "range to delete with scheduled false; range={:?}",
+                                            r,
+                                        );
+                                    }
+                                    true
+                                } else {
+                                    false
+                                }
```
Comment (overvenus) [inline components/range_cache_memory_engine/src/background.rs:1173]:
```suggestion
                                if range_being_delete == &r && !scheduled {
                                    panic!(
                                        "range to delete with scheduled false; range={:?}",
                                        r,
                                    );
                                };
                                range_being_delete == &r
```

Let's reduce indents.

File: components/range_cache_memory_engine/src/range_manager.rs [hunk 2]
```diff
@@ -479,6 +482,19 @@ impl RangeManager {
     pub fn get_and_reset_range_evictions(&self) -> u64 {
         self.range_evictions.swap(0, Ordering::Relaxed)
     }
+
+    // Only ranges that have not been scheduled will be retained in `ranges`
+    pub fn schedule_ranges_to_delete(&mut self, ranges: &mut Vec<CacheRange>) {
```
Comment (overvenus) [inline components/range_cache_memory_engine/src/range_manager.rs:487]:
```suggestion
    pub fn mark_delete_ranges_scheduled(&mut self, ranges: &mut Vec<CacheRange>) {
```

It does not schedule ranges but rather marks them as scheduled.

File: components/range_cache_memory_engine/src/range_manager.rs [hunk 3]
```diff
@@ -479,6 +482,19 @@ impl RangeManager {
...
+        ranges.retain(|r| {
+            let scheduled = self.ranges_being_deleted.get_mut(r).unwrap();
+            if !*scheduled {
+                *scheduled = true;
+                true
+            } else {
+                false
+            }
```
Comment (overvenus) [inline components/range_cache_memory_engine/src/range_manager.rs:495]:
```suggestion
            let has_scheduled = *scheduled;
            *scheduled = true;
            !has_scheduled
```

Comment (overvenus) [inline components/range_cache_memory_engine/src/range_manager.rs:495]:
```suggestion
            let has_scheduled = *scheduled;
            *scheduled = true;
            !has_scheduled
```

File: components/range_cache_memory_engine/src/engine.rs [hunk 4]
```diff
@@ -369,7 +369,9 @@ impl RangeCacheMemoryEngine {
     /// immediately due to some ongoing snapshots.
     pub fn evict_range(&self, range: &CacheRange) {
         let mut core = self.core.write();
-        let ranges_to_delete = core.range_manager.evict_range(range);
+        let mut ranges_to_delete = core.range_manager.evict_range(range);
+        core.mut_range_manager()
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/range_cache_memory_engine/src/engine.rs:373]: why remove mut_

File: components/range_cache_memory_engine/src/background.rs [hunk 5]
```diff
@@ -613,6 +613,10 @@ impl BackgroundRunnerCore {
                 range_stats_manager.handle_range_evicted(range);
             }
         }
+        self.engine
+            .write()
+            .mut_range_manager()
+            .schedule_ranges_to_delete(&mut ranges_to_delete);
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/range_cache_memory_engine/src/background.rs:619]: This line mark the ranges to be scheduled. Line 623 schedule them to the delete range worker.

**Review by overvenus — APPROVED** [review]
LGTM

**Review by LykxSassinator — APPROVED** [review]

**Pull request merged**